api.js

```diff
@@ -224,3 +236,86 @@ Api.prototype._run = function (files, _options) {
 		return runStatus;
 	});
 };
+
+Api._blankResults = {
+	stats: {
+		testCount: 0,
+		passCount: 0,
+		skipCount: 0,
+		todoCount: 0,
+		failCount: 0
+	},
+	tests: []
+};
+
+Api.prototype._runLimitedPool = function (files, runStatus, concurrency) {
+	var self = this;
+	var tests = {};
+
+	runStatus.on('timeout', function () {
+		Object.keys(tests).forEach(function (file) {
+			var fork = tests[file];
+			fork.exit();
+		});
+	});
+
+	return Promise.map(files, function (file) {
+		try {
+			var test = tests[file] = self._runFile(file, runStatus);
+
+			return new Promise(function (resolve) {
+				var runner = self._generateRunner(file, test, runStatus, resolve);
+				test.on('stats', runner);
+				test.on('exit', function () {
+					delete tests[file];
+				});
+				test.catch(runner);
+			});
+		} catch (err) {
+			runStatus.handleExceptions({
+				exception: err,
+				file: path.relative('.', file)
+			});
+
+			return Api._blankResults;
+		}
+	}, {concurrency: concurrency})
+		.then(function (results) {
+			// cancel debounced _onTimeout() from firing
+			if (self.options.timeout) {
+				runStatus._restartTimer.cancel();
+			}
+
+			runStatus.processResults(results);
+			return runStatus;
+		});
+};
+
+Api.prototype._generateRunner = function (file, test, runStatus, cb) {
```

Comment:
I renamed it cb but the the cb will always be the `resolve` from the promise above on line 268. 

`_generateRunner` is creating the callback for the `test.on('stats', ...)` event and this generated callback is what should eventually resolve the outer promise that represents the process in the concurrency pool. 

A good chunk of this is still a pull over from the old runner since I don't feel confident enough on the `self._runFile` logic and the fork objects to refactor so I just mimicked old behavior.